File: src/main/java/org/dvsa/testing/framework/jsoup/SpiderCrawler.java
package org.dvsa.testing.framework.jsoup;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.dvsa.testing.framework.axe.AXEScanner;
import org.dvsa.testing.framework.utils.DomainValidator;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import com.microsoft.playwright.Page;
import com.microsoft.playwright.options.LoadState;
import com.microsoft.playwright.options.WaitUntilState;
import org.openqa.selenium.WebDriver;

import java.net.URI;
import java.util.*;
import java.util.regex.Pattern;


public class SpiderCrawler {

    private static final Logger LOGGER = LogManager.getLogger(SpiderCrawler.class);
    private static final int MAX_CRAWL_DEPTH = 10;
    private static final int MAX_URLS_PER_DOMAIN = 1000;

    private static final Pattern EXCLUDE_PATTERN = Pattern.compile(
            ".*\\.(pdf|zip|gz|tar|mp3|mp4|avi|png|jpg|jpeg|gif|css|js|csv|docx|xlsx|exe|dmg)$",
            Pattern.CASE_INSENSITIVE
    );

    private static final Pattern ACTION_PATTERN = Pattern.compile(
            ".*(logout|signout|delete|remove|cancel|auth|login).*",
            Pattern.CASE_INSENSITIVE
    );


    public static void crawler(int level, String url, Set<String> visited, Object driver) {
        String baseDomain = DomainValidator.extractDomain(url);
        if (baseDomain == null) {
            LOGGER.error("Cannot extract domain from starting URL: {}", url);
            return;
        }
        crawlerWithDomain(level, url, visited, driver, baseDomain, false);
    }

    public static void crawlerWithDomain(int level, String url, Set<String> visited, Object driver, String baseDomain, boolean allowSubdomains) {
        if (level >= MAX_CRAWL_DEPTH || visited.size() >= MAX_URLS_PER_DOMAIN) {
            return;
        }

        String normalisedUrl = normaliseUrl(url);

        if (!DomainValidator.isSameDomain(normalisedUrl, baseDomain, allowSubdomains) || !visited.add(normalisedUrl)) {
            return;
        }

        LOGGER.info("Crawling [Level {}]: {}", level, normalisedUrl);

        try {
            String pageContent;
            String actualUrl;

            if (driver instanceof Page page) {
                page.navigate(normalisedUrl, new Page.NavigateOptions().setWaitUntil(WaitUntilState.LOAD));
                page.waitForLoadState(LoadState.DOMCONTENTLOADED);

                page.evaluate("() => { document.querySelectorAll('a[target]').forEach(l => l.removeAttribute('target')); }");

                actualUrl = page.url();
                pageContent = page.content();

            } else if (driver instanceof WebDriver seleniumDriver) {
                seleniumDriver.get(normalisedUrl);
                actualUrl = seleniumDriver.getCurrentUrl();
                pageContent = seleniumDriver.getPageSource();
            } else {
                throw new IllegalArgumentException("Unsupported driver type provided to crawler");
            }

            if (!DomainValidator.isSameDomain(actualUrl, baseDomain, allowSubdomains)) {
                LOGGER.debug("Redirected outside domain scope: {}", actualUrl);
                return;
            }

            AXEScanner.scan(driver);

            assert pageContent != null;
            assert actualUrl != null;
            Document doc = Jsoup.parse(pageContent, actualUrl);
            var links = doc.select("a[href]");

            for (Element link : links) {
                String nextUrl = link.absUrl("href");
                if (shouldFollow(nextUrl, baseDomain, visited, allowSubdomains)) {
                    crawlerWithDomain(level + 1, nextUrl, visited, driver, baseDomain, allowSubdomains);
                }
            }

        } catch (Exception e) {
            LOGGER.error("Failed to crawl URL [Level {}]: {}", level, normalisedUrl, e);
        }
    }

    private static String normaliseUrl(String urlString) {
        if (urlString == null || urlString.trim().isEmpty()) return "";
        try {
            URI uri = new URI(urlString.trim());
            String normalized = new URI(
                    uri.getScheme(),
                    uri.getAuthority(),
                    uri.getPath() != null ? uri.getPath() : "/",
                    null, null
            ).toString().toLowerCase();

            return normalized.endsWith("/") && normalized.length() > 1
                    ? normalized.substring(0, normalized.length() - 1)
                    : normalized;
        } catch (Exception e) {
            return urlString.toLowerCase().trim();
        }
    }

    private static boolean shouldFollow(String url, String baseDomain, Set<String> visited, boolean allowSubdomains) {
        if (url == null || url.trim().isEmpty() || !url.startsWith("http")) return false;
        if (!DomainValidator.isSameDomain(url, baseDomain, allowSubdomains)) return false;

        String normalized = normaliseUrl(url);
        if (visited.contains(normalized)) return false;

        String pathOnly = normalized.split("\\?")[0];
        if (EXCLUDE_PATTERN.matcher(pathOnly).matches()) return false;

        return !ACTION_PATTERN.matcher(normalized).matches();
    }
}

File: typescript/src/crawler/spiderCrawler.ts
import * as cheerio from 'cheerio';
import type { Page } from 'playwright';
import { AxeScanner } from '../axe/axeScanner.js';
import { logger } from '../logger.js';
import { isPlaywrightPage, type Driver } from '../types.js';
import { DomainValidator } from '../utils/domainValidator.js';

/** Port of org.dvsa.testing.framework.jsoup.SpiderCrawler (cheerio replaces JSoup). */

const MAX_CRAWL_DEPTH = 30;
const MAX_URLS_PER_DOMAIN = 1000;

const BLACKLISTED_PATHS = (process.env.SCANNER_EXCLUDE_PATHS ?? '/topsreport').split(',');

const EXCLUDE_PATTERN =
  /.*\.(pdf|zip|gz|tar|mp[34]|avi|png|jpg|jpeg|gif|svg|ico|css|js|map|csv|docx?|xlsx?|exe|dmg|woff2?|ttf|eot|json|xml)(\?.*)?$/i;

const ACTION_PATTERN = /.*(logout|signout|delete|remove|cancel|auth|login|abort).*/i;

function isInvalidUrl(url: string | null | undefined): boolean {
  if (!url || !url.trim() || url.startsWith('about:')) {
    return true;
  }

  if (ACTION_PATTERN.test(url)) {
    logger.warn(`Skipping potential state-changing URL: ${url}`);
    return true;
  }

  const isBlacklisted = BLACKLISTED_PATHS.some((path) =>
    url.toLowerCase().includes(path.toLowerCase()),
  );
  if (isBlacklisted) {
    logger.info(`URL ignored (Blacklisted path): ${url}`);
    return true;
  }

  const pathOnly = url.split('?')[0];
  return EXCLUDE_PATTERN.test(pathOnly);
}

export function normaliseUrl(urlString: string | null | undefined): string {
  if (!urlString || !urlString.trim()) return '';

  try {
    const uri = new URL(urlString.trim());
    let normalized = `${uri.protocol}//${uri.host}${uri.pathname || '/'}`.toLowerCase();

    if (normalized.endsWith('/') && normalized.length > 8) {
      normalized = normalized.slice(0, -1);
    }
    return normalized;
  } catch {
    let fallback = urlString.trim().toLowerCase();
    if (fallback.includes('#')) fallback = fallback.split('#')[0];
    if (fallback.includes('?')) fallback = fallback.split('?')[0];
    if (fallback.endsWith('/') && fallback.length > 8) {
      fallback = fallback.slice(0, -1);
    }
    return fallback;
  }
}

function shouldFollow(
  url: string,
  baseDomain: string,
  visited: Set<string>,
  allowSubdomains: boolean,
): boolean {
  if (isInvalidUrl(url)) return false;
  if (!DomainValidator.isSameDomain(url, baseDomain, allowSubdomains)) return false;
  return !visited.has(normaliseUrl(url));
}

async function getPageSource(driver: Driver): Promise<string> {
  if (isPlaywrightPage(driver)) {
    return driver.content();
  }
  return driver.getPageSource();
}

export async function crawler(
  level: number,
  url: string,
  visited: Set<string>,
  driver: Driver,
): Promise<void> {
  const baseDomain = DomainValidator.extractDomain(url);
  if (!baseDomain) {
    logger.error(`Cannot extract domain from starting URL: ${url}`);
    return;
  }
  await crawlerWithDomain(level, url, visited, driver, baseDomain, false);
}

export async function crawlerWithDomain(
  level: number,
  url: string,
  visited: Set<string>,
  driver: Driver,
  baseDomain: string,
  allowSubdomains: boolean,
): Promise<void> {
  if (level >= MAX_CRAWL_DEPTH || visited.size >= MAX_URLS_PER_DOMAIN) {
    return;
  }

  const normalisedUrl = normaliseUrl(url);

  if (isInvalidUrl(normalisedUrl)) {
    return;
  }

  if (visited.has(normalisedUrl)) {
    logger.debug(`Skipping: Base URL already visited -> ${normalisedUrl}`);
    return;
  }
  visited.add(normalisedUrl);

  if (!DomainValidator.isSameDomain(normalisedUrl, baseDomain, allowSubdomains)) {
    return;
  }

  logger.info(`Crawling [Level ${level}]: ${normalisedUrl}`);

  try {
    let actualUrl: string;
    if (isPlaywrightPage(driver)) {
      const page: Page = driver;
      await page.goto(normalisedUrl, { waitUntil: 'load' });
      await page.waitForLoadState('networkidle');
      await page.evaluate(() => {
        document.querySelectorAll('a[target]').forEach((link) => link.removeAttribute('target'));
      });
      actualUrl = page.url();
    } else {
      await driver.get(normalisedUrl);
      actualUrl = await driver.getCurrentUrl();
    }

    if (!DomainValidator.isSameDomain(actualUrl, baseDomain, allowSubdomains)) {
      logger.debug(`Redirected outside domain: ${actualUrl}`);
      return;
    }

    await AxeScanner.scan(driver);

    const freshContent = await getPageSource(driver);
    const $ = cheerio.load(freshContent);

    const links: string[] = [];
    $('a[href]').each((_, element) => {
      const href = $(element).attr('href');
      if (!href) return;
      try {
        links.push(new URL(href, actualUrl).toString());
      } catch {
        /* unresolvable href — skip */
      }
    });

    for (const nextUrl of links) {
      if (shouldFollow(nextUrl, baseDomain, visited, allowSubdomains)) {
        await crawlerWithDomain(level + 1, nextUrl, visited, driver, baseDomain, allowSubdomains);
      }
    }
  } catch (e) {
    logger.error(`Failed to crawl URL [Level ${level}]: ${normalisedUrl}`, e);
  }
}
